src/backend/distributed/transaction/remote_transaction.c

```diff
@@ -160,6 +169,74 @@ BeginAndSetDistributedTransactionIdCommand(void)
 }
 
 
+/*
+ * BeginTransactionCommand returns the BEGIN command to use for the current isolation
+ * level.
+ *
+ * Transactions have 3 properties that we care about here:
+ * - XactIsoLevel (isolation level)
+ * - XactDeferrable (deferrable)
+ * - XactReadOnly (read only)
+ *
+ * These properties can be set in several ways:
+ * - via BEGIN TRANSACTION ISOLATION LEVEL ...
+ * - via default_transaction_isolation, ...
+ * - via SET TRANSACTION .. (or plain SET transaction_isolation ...)
+ *
+ * We want to make sure that the properties that are passed to the worker nodes
+ * match the coordinator as much as possible. However, we do not want to waste
+ * bytes repeating the current values ad infinitum.
+ *
+ * The trade-off we make is that we send the isolation level in all cases,
+ * but only set deferrable and read-only if they were explicitly specified
+ * in the BEGIN by the user. The implication is that we may not follow the
+ * default_transaction_* settings on the coordinator if they differ on the
+ * worker.
+ */
+static char *
+BeginTransactionCommand(void)
+{
+	StringInfo beginCommand = makeStringInfo();
+
+	/*
+	 * XactIsoLevel can only be set at the start of the transaction, before the
+	 * first query. Since Citus does not send BEGIN until the first query, we
+	 * can simply use the current values, and they will match the values for the
+	 * outer transaction after any BEGIN and SET TRANSACTION that may have occurred.
+	 */
+	appendStringInfo(beginCommand, "BEGIN TRANSACTION ISOLATION LEVEL %s",
+					 IsolationLevelName[XactIsoLevel]);
+
+	if (HasBeginXactDeferrable)
+	{
+		if (BeginXactDeferrable)
+		{
+			appendStringInfoString(beginCommand, " DEFERRABLE");
+		}
+		else
+		{
+			appendStringInfoString(beginCommand, " NOT DEFERRABLE");
+		}
+	}
+
+	if (HasBeginXactReadOnly)
```

Comment:
> A slight challenge with the read-only behaviour is that it can be changed by subtransactions if the outer transaction is read-write. Therefore, we cannot simply look at the current XactReadOnly, which might reflect the value within the subtransaction that was set using SET TRANSACTION. We cannot easily know the original value, since it is captured in the GUC stack. We therefore add a small hook to capture explicit BEGIN properties from the utility hook.

Still, shouldn't we respect `XactReadOnly` at the time when we call `BeginTransactionCommand`? We might be on a subtx, and if the subtx sets to read-only mode, shouldn't new remote connections/transactions get the current setting? 

I think if the user does `SET LOCAL transaction_read_only to XXX` we are already sending it to the existing remote sub transactions/connections.

For example, for the SELECT query below, I'd expected that we send `transaction_read_only TO YES;`.  What am I missing here :D

```SQL
local] onderkalaci@postgres:5432-79558=# BEGIN;
BEGIN
Time: 0.345 ms
[local] onderkalaci@postgres:5432-79558=# insert into distributed_table_cdtc VALUES (1);
NOTICE:  issuing BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;SELECT assign_distributed_transaction_id(0, 33, '2022-11-09 15:57:11.414518+01');
DETAIL:  on server onderkalaci@localhost:9701 connectionId: 2
NOTICE:  issuing INSERT INTO public.distributed_table_cdtc_102010 (column1) VALUES (1)
DETAIL:  on server onderkalaci@localhost:9701 connectionId: 2
INSERT 0 1
Time: 4.382 ms

[local] onderkalaci@postgres:5432-79558=# SAVEPOINT s1;
NOTICE:  issuing SAVEPOINT savepoint_2
DETAIL:  on server onderkalaci@localhost:9701 connectionId: 2
SAVEPOINT
Time: 1.564 ms
[local] onderkalaci@postgres:5432-79558=# SET transaction_read_only TO YES;
SET
Time: 1.089 ms
[local] onderkalaci@postgres:5432-79558=# SELECT * from distributed_table_cdtc WHERE column1 = 2;
NOTICE:  issuing BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;SAVEPOINT savepoint_2;SELECT assign_distributed_transaction_id(0, 33, '2022-11-09 15:57:11.414518+01');
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
NOTICE:  issuing SELECT column1 FROM public.distributed_table_cdtc_102033 distributed_table_cdtc WHERE (column1 OPERATOR(pg_catalog.=) 2)
DETAIL:  on server onderkalaci@localhost:9700 connectionId: 1
┌─────────┐
│ column1 │
├─────────┤
└─────────┘
(0 rows)

Time: 5.952 ms
````  

If I try to insert, I'd get the error from the coordinator's `standard_ExecutorStart`, so hard to find a hole to get into a state where we behave unexpectedly anyway

Reply:
> We might be on a subtx, and if the subtx sets to read-only mode, shouldn't new remote connections/transactions get the current setting?

no, because that remote transaction cannot be reset to read-write, but the outer transaction can.

>  What am I missing here :D

we don't propagate SET :-/

Reply:
Alright, I think I see your point now. We already send all the `SET LOCAL` ones that happens in between, so basically by replicating the BEGIN how user typed, we can practically have the same settings as the coordinator setting on each (sub)transaction level.

The big exception here is SET commands.

We might add a test for this (if we don't have already one) -- assuming value 2 is already in the table:

```SQL
[local] onderkalaci@postgres:5432-49525=# set citus.propagate_set_commands to local ;
SET
Time: 0.395 ms
[local] onderkalaci@postgres:5432-49525=# BEIN;
ERROR:  syntax error at or near "BEIN"
LINE 1: BEIN;
        ^
Time: 0.316 ms
[local] onderkalaci@postgres:5432-49525=# BEGIN;
BEGIN
Time: 0.376 ms
[local] onderkalaci@postgres:5432-49525=#  insert into distributed_table_cdtc VALUES (1); 
INSERT 0 1
Time: 1.901 ms
[local] onderkalaci@postgres:5432-49525=# SAVEPOINT s1;
SAVEPOINT
Time: 0.984 ms
[local] onderkalaci@postgres:5432-49525=# SET LOCAL transaction_read_only TO YES;        SET
Time: 1.078 ms
[local] onderkalaci@postgres:5432-49525=#  SELECT current_setting('transaction_read_only') from distributed_table_cdtc WHERE column1 = 2;
┌─────────────────┐
│ current_setting │
├─────────────────┤
│ on              │
└─────────────────┘
(1 row)

 SELECT current_setting('transaction_read_only') from distributed_table_cdtc WHERE column1 = 1;
┌─────────────────┐
│ current_setting │
├─────────────────┤
│ on              │
└─────────────────┘
(1 row)


rollback to savepoint s1;
ROLLBACK
Time: 10.412 ms
[local] onderkalaci@postgres:5432-49525=#  SELECT current_setting('transaction_read_only') from distributed_table_cdtc WHERE column1 = 1;
┌─────────────────┐
│ current_setting │
├─────────────────┤
│ off             │
└─────────────────┘
(1 row)
SELECT current_setting('transaction_read_only') from distributed_table_cdtc WHERE column1 = 2;
┌─────────────────┐
│ current_setting │
├─────────────────┤
│ off             │
└─────────────────┘
(1 row)


```

Reply:
Also, it was a little bit counter initiative to see this code, as IsolationLevelName and others are handled differently. Might be good to move some of the PR description to here (or another place in the code?)

Reply:
The function comment has details on the trade-offs.

I mainly wanted to avoid adding DEFERRABLE or READ WRITE to every BEGIN. 

We could perhaps consider omit isolation level too. Probably a bit painful in terms of (failure) test changes.